### ISSUE TITLE
Redirection issue when using Angular.js

### ISSUE THREAD

**Issue by GioPan** [jaredhanson/passport-facebook#106]
I was hoping you might have some suggestion for this question on Stackoverflow: 
http://stackoverflow.com/questions/27998501/passport-js-express-js-and-angular-js-routing-how-can-they-coexist . 
Summary: I am using Passport.js authorization. When Facebook authorizes access, I cannot redirect to a specific page from Express.js since the routing is intercepted by Angular.js routing.
I hope you can help, this is driving me crazy.

**Comment by johnsonw** [comment]
Try the following:

1. On the link, do something like this:
<a href="/auth/facebook?returnTo={{location}}">Sign in with Facebook</a>

where {{location}} is the value of $location.url();

Then, in your express router:

 // Redirect the user to Facebook for authentication.  When complete, Facebook
      // will redirect the user back to the application at
      //   /auth/facebook/callback
      .get ('/auth/facebook', function authenticateFacebook (req, res, next) {
        session.returnTo = '/#' + req.query.returnTo; 
        next ();
      },
      passport.authenticate ('facebook')
    )
      .get ('/auth/facebook/callback', function (req, res, next) {
      var authenticator = passport.authenticate ('facebook', {
        successRedirect: session.returnTo,
        failureRedirect: '/login'
      });

      delete session.returnTo;
      authenticator (req, res, next);
    })

**Comment by GioPan** [comment]
Thanks. I tried what you suggested but any redirection of type 
`successRedirect:/#/successRoute` and `failureRedirect:/#/failureRoute`
end up in `domain/#_=_`, so in the home view.

**Comment by Fonger** [comment]
@GioPan I solve the suffix #_=_ by the following code on client site javascript.

Just push the following code **before** angularjs and it works for me.

```js
// Remove the ugly Facebook appended hash
// <https://github.com/jaredhanson/passport-facebook/issues/12>
if (window.location.hash && window.location.hash === "#_=_") {
	if (window.history && history.pushState) {
		window.history.pushState("", document.title, window.location.pathname);
	} else {
	// Prevent scrolling by storing the page's current scroll offset
	var _scroll = {
		top: document.body.scrollTop,
		left: document.body.scrollLeft
	};
	window.location.hash = "";
	// Restore the scroll offset, should be flicker free
	document.body.scrollTop = _scroll.top;
	document.body.scrollLeft = _scroll.left;
	}
}
```

**Comment by jerry153fish** [comment]
Thanks

**Comment by appyapp** [comment]
Hi everyone,

I tried everything you said but I still got the problem.
I use the $routeProvider in my angular front-end:

```
  $routeProvider.
      when('/', {
        templateUrl: 'veamus/menu.html',
        controller: 'IndexCtrl'
      }).
      when('/emisor', {
        templateUrl: 'veamus/emisor.html',
        controller: 'EmisorCtrl',
        controllerAs: 'controller'
      }).
      when('/receptor', {
        templateUrl: 'veamus/receptor.html',
        controller: 'ReceptorCtrl',
        controllerAs: 'controller'
      }).
      when('/login',{
        templateUrl: 'veamus/login.html',
        controller: 'LoginCtrl'
      }).
      when('/signup',{
        templateUrl: 'veamus/signup.html',
        controller: 'SignupCtrl'
      }).
      otherwise({
        redirectTo: '/'
      });
    $locationProvider.html5Mode(true);
```
The facebook callback URL ( in my case /auth/facebook/callback) is not present in my routes. At the redirection, I come back in the Angular Router and as he doesn't know the route, I'm redirected at the index page (url : /).
I tried without the otherwise and logically I got a blank template as '/auth/facebook/callback' is a unknown route.

In the server-side, I got all my routes well defined according to the documentation :
```
//Facebook Login Routes
app.get('/auth/facebook',passport.authenticate('facebook',{scope: 'email' } );
app.get('auth/facebook/callback',passport.authenticate('facebook',{
failureRedirect : '/login',
successRedirect : '/'
}));
```

In my server app, I changed the failure and success redirects URL to '/#/login' but with no luck.

So, basically, I can access to the Facebook login page (thanks to target="_self" in the link attributes) but the redirection ends in my Angular router and not in the server-side.

How can I resolve this issue ?

FYI, I tried to create a new route in my angular routeProvider as follow :
```
   when('/auth/facebook/callback',{
        redirectTo: function(skip,url){
          window.location.href = 'http://localhost:8031/auth/facebook/callback';
        }
      })
```

It ends up in infinite loop...

**Comment by robsonfagundes** [comment]
Hi there!

Folks I have the same problem, I'd like that after authentication in my back-end app, I need call the route wishes in front-end.

Summary: I have two apps separated: 
API: (NodeJS, Express, MongoDB in back-end) and APP: (AngularJS front-end). 

my code:
`
app.get('/auth/github/callback',
	passport.authenticate('github', {
		successRedirect: '/#/successRoute',
		failureRedirect: '/auth/github',
	})
);
`

Someone knows how I do it. Thankyou!

**Comment by labriffa** [comment]
I tried something similar to @johnsonw but I still couldn't quite get this to work. But with a bit of tweaking I managed to get it routing how I wanted.

The configuration I used was

```
module.exports = {
    'twitter': {
        'clientID': process.env.TWITTER_KEY,
        'clientSecret': process.env.TWITTER_SECRET,
        'callbackURL': process.env.APP_URL + 'auth/twitter/callback?returnTo="/"'
    }
};
```
My Twitter callback was

```
app.route('/auth/twitter/callback')
        .get(passport.authenticate('twitter', {
           successRedirect: '/',
           failureRedirect: '/login'
        }));
```

My Angular App currently redirects to the /login page whenever it makes unauthorized API calls to the Express backend via $http interceptors.

```
app.factory('loginObserver', function loginObserver($q, $location) {
    return {
        'responseError': function(response) {
            var status = response.status;
            switch (status) {
                case 403:
                    $location.path("/login");
                    return response;
                    break;
                default:
                    return $q.reject(response);
            }
        }
    }
});
```

For every other route that makes use of my middleware **_isLoggedIn()_** I check to see if it's an API request in which case if it is I return the 403 otherwise I delegate AngularJS with the responsbility of routing to the login page.
```
function isLoggedIn(req, res, next) {
        if(req.isAuthenticated()) {
            return next();
        } else {
            if(req.url.indexOf('api')===-1) {
                res.redirect('/#/login');
            } else {
                res.writeHead(403);
                res.end();
            }
        }
    }
```

**Comment by yogeshbansal** [comment]
@robsonfagundes: Are you able to solve your issue

**Comment by yogeshbansal** [comment]
I am also having two apps separately, angular 2 as front end running on port 4200 and node express MongoDB app as backend at port 3000. If I hit the /auth/facbook URL directly from the browser, it working fine, if going to Facebook login page. But when I hit it from angular 2 front end app. I am getting this error: 
login:1 Failed to load https://www.facebook.com/dialog/oauth?response_type=code&redirect_uri=https%3A%2F%2Fapi.kharidaree.com%3A8443%2Fauth%2Ffacebook%2Fcallback&scope=email%2Cpublic_profile&client_id=399438350415575: Redirect from 'https://www.facebook.com/dialog/oauth?response_type=code&redirect_uri=https%3A%2F%2Fapi.kharidaree.com%3A8443%2Fauth%2Ffacebook%2Fcallback&scope=email%2Cpublic_profile&client_id=399438350415575' to 'https://www.facebook.com/login.php?skip_api_login=1&api_key=399438350415575&signed_next=1&next=https%3A%2F%2Fwww.facebook.com%2Fv2.8%2Fdialog%2Foauth%3Fredirect_uri%3Dhttps%253A%252F%252Fapi.kharidaree.com%253A8443%252Fauth%252Ffacebook%252Fcallback%26scope%3Demail%252Cpublic_profile%26response_type%3Dcode%26client_id%3D399438350415575%26ret%3Dlogin%26logger_id%3D8670ab73-50a9-295d-4f84-826480f7efff&cancel_url=https%3A%2F%2Fapi.kharidaree.com%3A8443%2Fauth%2Ffacebook%2Fcallback%3Ferror%3Daccess_denied%26error_code%3D200%26error_description%3DPermissions%2Berror%26error_reason%3Duser_denied%23_%3D_&display=page&locale=en_US&logger_id=8670ab73-50a9-295d-4f84-826480f7efff' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'null' is therefore not allowed access.

**Comment by robsonfagundes** [comment]
@yogeshbansal I did not solve until now...

**Comment by karansethi** [comment]
Is This issue fixed? I am also facing the exact same issue

**Comment by abhishek-b-agarwal** [comment]
Hey @yogeshbansal,

Were you able to solve the issue you mentioned above?

Abi
